### PR TITLE
Enable abort job functionality.

### PR DESCRIPTION
**Pull request by carolinar7**
This PR adds the abort feature to the backend. Future work will be done on the UI side, to enable the user to abort jobs while they are in a Triggered or Pending state in Prow's UI.

Fixes: https://github.com/kubernetes/test-infra/issues/24390

/cc @chaodaiG @cjwagner

### CODE REVIEW

File: prow/cmd/deck/abort.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (chaodaiG) [inline prow/cmd/deck/abort.go:2]: 2022

File: prow/cmd/deck/abort.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+				// admins only care about errors other than not found
+				l.WithError(err).Warning("ProwJob not found.")
+			}
+			return
+		}
+		switch r.Method {
+		case http.MethodPost:
+			if pj.Status.State != prowapi.TriggeredState && pj.Status.State != prowapi.PendingState {
+				http.Error(w, fmt.Sprintf("Cannot abort job with state: %q", pj.Status.State), http.StatusBadRequest)
+				l.Error("Cannot abort job with state")
```
Comment (chaodaiG) [inline prow/cmd/deck/abort.go:57]: normally we log error or warning when the error requires prow operator's attention, particularly due to potential bug in prow. This is more like user error or race condition, let's log as Debug instead

File: prow/cmd/deck/abort.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			allowed, err, code := isAllowedToRerun(r, cfg, goa, ghc, *pj, cli, pluginAgent, l)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("Could not verify if allowed to abort: %v", err), code)
+				l.WithError(err).Debug("Could not verify if allowed to abort")
+			}
+			l = l.WithField("allowed", allowed)
+			l.Info("Attempted abort")
+			if !allowed {
+				http.Error(w, "You don't have permission to abort this job", http.StatusUnauthorized)
+				l.Error("Error writing to abort response.")
```
Comment (chaodaiG) [inline prow/cmd/deck/abort.go:71]: this is the wrong error message, also this should also be `Debug`

File: prow/cmd/deck/abort.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+				http.Error(w, fmt.Sprintf("Error marshal source job: %v", err), http.StatusInternalServerError)
+				l.WithError(err).Errorf("Error marshal source job")
+			}
+			pj, err := prowJobClient.Patch(ctx, pj.Name, ktypes.MergePatchType, jsonPJ, metav1.PatchOptions{})
+			if err != nil {
+				http.Error(w, fmt.Sprintf("Could not patch aborted job: %v", err), http.StatusInternalServerError)
+				l.WithError(err).Errorf("Could not patch aborted job")
+			}
+			l.Info("Successfully aborted PJ.")
+			json.NewEncoder(w).Encode(pj)
```
Comment (chaodaiG) [inline prow/cmd/deck/abort.go:87]: what does this line do?

File: prow/cmd/deck/abort.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			if _, err = w.Write([]byte("Job successfully aborted.")); err != nil {
+				l.WithError(err).Error("Error writing to abort response.")
```
Comment (chaodaiG) [inline prow/cmd/deck/abort.go:89]: I think it's fine for this to fail, user will see that the job is aborted from the UI almost instantly, let's log this `Debug` as well

File: prow/cmd/deck/abort.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+			}
+			l = l.WithField("allowed", allowed)
+			l.Info("Attempted abort")
+			if !allowed {
+				http.Error(w, "You don't have permission to abort this job", http.StatusUnauthorized)
+				l.Error("Error writing to abort response.")
+				return
+			}
+			pj.SetComplete()
+			pj.Status.State = prowapi.AbortedState
```
Comment (carolinar7) [inline prow/cmd/deck/abort.go:75]: Add pj.Status.Description here indicating who aborted the job

File: prow/cmd/deck/abort.go (hunk 1 above)
Comment (carolinar7) [reply prow/cmd/deck/abort.go]: updated!

File: prow/cmd/deck/abort.go (hunk 2 above)
Comment (carolinar7) [reply prow/cmd/deck/abort.go]: updated!

File: prow/cmd/deck/abort.go (hunk 3 above)
Comment (carolinar7) [reply prow/cmd/deck/abort.go]: What do you mean by "this is the wrong error message"? Also, Error has been updated to Debug!

File: prow/cmd/deck/abort.go (hunk 4 above)
Comment (carolinar7) [reply prow/cmd/deck/abort.go:94]: `json.NewEncoder(w).Encode(pj)` is basically marshalling the prow job, and writing it as a body response to the post request. I added a comment to make this clear.

File: prow/cmd/deck/abort.go (hunk 5 above)
Comment (carolinar7) [reply prow/cmd/deck/abort.go]: Updated!

File: prow/cmd/deck/abort.go (hunk 6 above)
Comment (carolinar7) [reply prow/cmd/deck/abort.go:80]: Updated!

File: prow/cmd/deck/abort.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+				l.WithError(err).Debug("Could not verify if allowed to abort")
+			}
+			l = l.WithField("allowed", allowed)
+			l.Info("Attempted abort")
+			if !allowed {
+				http.Error(w, "You don't have permission to abort this job", http.StatusUnauthorized)
+				l.Error("Error writing to abort response.")
+				return
+			}
+			pj.SetComplete()
```
[the remark this answers is not in the document]
Comment (carolinar7) [reply prow/cmd/deck/abort.go]: Do you know if plank will eventually pick up this PJ, and set the complete time? In this case, what do you believe should be tested in unit testing. Is it good enough just to check that the status of the job is aborted. Then, create an integration test that also checks that a complete time has been set by Plank?

File: prow/cmd/deck/abort_test.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+// TestAbort that an aborted job has an updated status and
+// that permissions were granted appropriately
+func TestAbort(t *testing.T) {
+	testCases := []struct {
+		name        string
+		login       string
+		authorized  []string
+		allowAnyone bool
+		jobState    prowapi.ProwJobState
```
Comment (chaodaiG) [inline prow/cmd/deck/abort_test.go:45]: the abort function also updates the description, which should also be tested other than job state

File: prow/cmd/deck/abort.go (hunk 3 above)
Comment (chaodaiG) [reply prow/cmd/deck/abort.go]: the error is "you don't have permission to abort", not "error writing to abort response"

File: prow/cmd/deck/abort.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+				http.Error(w, fmt.Sprintf("Error marshal source job: %v", err), http.StatusInternalServerError)
+				l.WithError(err).Errorf("Error marshal source job")
+			}
+			pj, err := prowJobClient.Patch(ctx, pj.Name, ktypes.MergePatchType, jsonPJ, metav1.PatchOptions{})
+			if err != nil {
+				http.Error(w, fmt.Sprintf("Could not patch aborted job: %v", err), http.StatusInternalServerError)
+				l.WithError(err).Errorf("Could not patch aborted job")
+			}
+			l.Info(abortDescription)
+			// Marshalls pj and adds it to body of response
```
Comment (chaodaiG) [inline prow/cmd/deck/abort.go:93]: does the consumer of the response expect anything from the prowjob struct other than the status code? if not I don't think we need to include the entire prowjob struct as body

File: prow/cmd/deck/abort.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (carolinar7) [reply prow/cmd/deck/abort.go]: Thanks, and updates to the unit tests have been made!

**Review by chaodaiG — APPROVED** [review]

**Pull request merged**